### ISSUE TITLE
alternate ternary formatting

### ISSUE THREAD

**Issue by mvolkmann** [prettier/prettier#737]
Here is how Prettier currently reformats nested ternaries:
````
const value = condition1
  ? value1
  : condition2
      ? value2
      : condition3
          ? value3
          : value4;
````
This isn't so bad when there are only two ternaries, but I sometimes have many more because I essentially use them as an alternative to a switch statement when each case only gets a single value.

Here is how I prefer to format this case:
````
const value =
  condition1 ? value1 :
  condition2 ? value2 :
  condition3 ? value3 :
  value4;
````
I like this because it clearly associates each condition with its corresponding value and the indentation doesn't increase regardless of the number of ternaries.

Perhaps some like the current formatting because it discourages use of nested ternaries, but I think my example shows that you can have any number without making the code hard to read.

Would you consider adding an option to support this alternate formatting? Maybe something like this? `--ternary=condval`

**Comment by ggoodman** [comment]
A minor tweak to the current ternary formatting that I favour is like this:

```js
const score = doc.thumbs
    ?   _.reduce(_.range(1, doc.thumbs), (score, n) => {
            return score + baseIncrement / Math.pow(decayFactor, n);
        }, doc.created_at.valueOf())
    :   doc.created_at.valueOf();
```

Specifically, I like to move to the next tab stop after the `?` and `:` portions of the operator. This means that multi-line 'values' are expressed at a consistent indent-level with other blocks.

As you can see above, the logic that is relevant to the situation where `doc.thumbs` is truthy is easy to identify by visual scanning.

As a reference point, here is the output of the same block of code via `prettier`:
```js
const score = doc.thumbs
    ? _.reduce(
          _.range(1, doc.thumbs),
          (score, n) => {
              return score + baseIncrement / Math.pow(decayFactor, n);
          },
          doc.created_at.valueOf()
      )
    : doc.created_at.valueOf();
```

In the latter, you can see that while the reducer function is at a consistent indentation level, it is out of sync with my configured 4-space indentation level (it is indented by 6 spaces). Further indentations within that block propagate this unusual level of indentation to 10 spaces, 14 spaces, etc...

**Comment by nfroidure** [comment]
I agree that nested ternaries do not need indentation since it is clear that they are just acting on the same output value.

The only thing we need is to be able to distinguish conditions from actual values which is clearly visible when prefixing with the `?`/`:` tokens, this is why i find this formatting is the most readable:

```
const value = condition1
  ? value1
  : condition2
  ? value2
  : condition3
  ? value3
  : value4;
```

**Comment by yamafaktory** [comment]
Somehow related to #716 (closed). Honestly, I would never ship production code that looks like the last comment (no offense, but this far from being readable).

**Comment by nfroidure** [comment]
No offense but no input either, except a feeling. Would have been nicer to elaborate on why you find it unreadable.

**Comment by yamafaktory** [comment]
Did you read the aforementioned similar closed issue? IMHO Prettier is clearly opinionated and I think @jlongster explained it really well. I also see formatting ternaries the way you did as abusing this syntax when there are other ways to do it.

**Comment by nfroidure** [comment]
I did but there is no mention of the above coding style. No problem, i just wanted to give some input on a potential usage of ternary operators.

**Comment by rhengles** [comment]
@mvolkmann Implemented in https://github.com/arijs/prettier-with-tabs/commit/290aee15e02de453909cdd8a7a91c3ecaf74bfad as `flattenTernaries` option

**Comment by mvolkmann** [comment]
Thanks Rafael!

---
R. Mark Volkmann
Object Computing, Inc.

**Comment by vjeux** [comment]
This keeps confusing people both inside and outside of Facebook. Here's a proposal:

If there is a strict chain of ternaries, then write it the following way:

```js
a
  ? b
  : c
  ? d
  : e
  ? f
  : g
```

We need to make sure that the fully nested ones are expanded, otherwise it looks really weird.

This is not as good looking as

```js
a ? b :
c ? d :
e ? f :
g
```

but at least it's compatible with the current way we print ternaries and it doesn't have issues when both don't fit in a single line.

What do you think?

**Comment by AshCoolman** [comment]
I use ternaries often to act as an "if else" expression within my JSX:

```jsx
const StorybookLoader = ({ match }) => (
  match.params.storyId === "button"
    ? <ButtonStorybook />
    : match.params.storyId === "color"
    ? <ColorBook />
    : match.params.storyId === "typography"
    ? <TypographyBook />
    : match.params.storyId === "loading"
    ? <LoaderStorybook />
    : match.params.storyId === "deal-list"
    ? <DealListStory />
    : (
      <Message>
        <Title>{'Missing story book'}</Title>
        <Content>
          <BackButton/>
        </Content>
      </Message>
    )
)
```

In my opinion, this is more readable. 

In all cases I can imagine, all nested ternaries look better when flat.

**Comment by MrEfrem** [comment]
Why there is no continuation here? Do how to fix the warning in `eslint`?

**Comment by lydell** [comment]
@MrEfrem Easy: Turn off ESLint rules that conflict with Prettier :)

**Comment by MrEfrem** [comment]
Then is need to turn off entirely rule "indent", what isn't good.

**Comment by lydell** [comment]
@MrEfrem I'm curious, why isn't good to turn off "indent"?

**Comment by MrEfrem** [comment]
Because there `http://eslint.org/docs/rules/indent` a lot of sub-rules and I don't want to turn off it. Not all in command to use `prettier` and turn off `indent` will be no good idea.

**Comment by lydell** [comment]
But if the indent rule reports any problems you can't fix them, can you? Because Prettier takes care of all indentation.

**Comment by MrEfrem** [comment]
I can to fix them only I should turn off `prettier` for it, to change the file and turn on `prettier` again for other files. I use `vscode` + `prettier plugin` and often have turning off it.

**Comment by Jamesernator** [comment]
Ternary indentation is always controversial, personally I write like more like `if-elseif-else` chains:

```js
const message =
    i % 3 === 0 && i % 5 === 0 ?
        'fizzbuzz'
    : i % 3 === 0 ?
        'fizz'
    : i % 5 === 0 ?
        'buzz'
    :
        String(i)
```

**Comment by MazeChaZer** [comment]
@Jamesernator I really like this idea. If you're having multiple branches in your ternary operator, chances are you are doing functional programming. And this looks just like an if-then-else-expressions in real functional languages like Elm. To me this is very readable.

**Comment by 0x24a537r9** [comment]
I think @Jamesernator is onto something with indenting the cases, and @MazeChaZer with their likening to functional if/else blocks. I do suspect putting the `:` at the beginning and `?` at the end might be a bit obscure for JS newcomers though, who might mistake it for some label-like syntax. In general, I think it's best to be consistent (at least within a statement) with whether binary operators go on the same line or the next (both are fine). The else case also leaves a lot to be desired.

While I've historically used @mvolkmann's proposal for its brevity, I've found in practice that its scannability quickly degenerates if the condition/value need to wrap a line (so it's not great as a general rule), and I've had experience with code reviewers newer to the language having trouble parsing the format with the postfix `:` when the `?` was infix. As such, I think it is probably wiser for any general rule to separate condition and value by line.

I think all of the proposals so far have some wisdom to them, so it would great if we could take the best parts of each. Might I propose a compromise between @mvolkmann's, @vjeux's, and @Jamesernator's proposals:
```
const value =
  condition1 ?
    value1 :
  condition2 ?
    value2 :
  condition3 ?
    value3 :
  value4;
```
This has a few advantages as I see it:
  - Conditions end with question marks, which coheres with English sentence structure. (like @mvolkmann's and @Jamesernator's)
  - Ternary punctuation is kept away from the beginning of the line, reducing visual noise that might make indentation difficult to scan (like @mvolkmann's)
  - Leverages if/switch-block-like indentation semantics for super clear distinction between conditions and values, while also improving quick scanning of either conditions or values. Very consistent with functional programming styles (which, is probably what you're going for if you're using nested ternaries, like @Jamesernator's)
  - Handles long expressions well when conditions or values need to wrap (like @vjeux's)
  - More consistent with existing (postfix) prettier rules that place binary operators (e.g. `&&`/`||`) at the ends of lines. Internally consistent as well.

It's hard to please everyone, but I think this achieves a fair balance of the key features of each of the proposals above while remaining approachable to the JS noob.

**Comment by JAForbes** [comment]
I've noticed people often vary in their preference for ternary formatting.  Just throwing in my favorite way for reference, but as long as we're avoiding nesting each new branch - I'm happy.

```js
const paymentMessage =
  state == 'success'
    ? 'Payment completed successfully'

  : state == 'processing'
    ? 'Payment processing'

  : state == 'invalid_cvc'
    ? 'There was an issue with your CVC number'

  : state == 'invalid_expiry'
    ? 'Expiry must be sometime in the past.'

    : 'There was an issue with the payment.  Please contact support.'
```

Which isn't so different to an if statement in an expression oriented language.  e.g.

```coffeescript


paymentMessage = 

  if state == 'success'
    'Payment completed successfully'

  else if (state == 'processing')
    'Payment processing'

  else if (state == 'invalid_cvc')
    'There was an issue with your CVC number'

  else if (state == 'invalid_expiry')
    'Expiry must be sometime in the past.'

  else 
    'There was an issue with the payment.  Please contact support.'

```

I'd be very happy with other suggestions too.

The key feature for me ( that I've also seen in above proposals) is a  line break and indent after the initial `=` so conditions and results all have the same nesting.  It makes it easier for me to scan.

I'd _prefer_ the operator was either all first, or all last, not alternating.  I prefer first because I often comment out segments and replace them temporarily while debugging.  e.g.

```js
const paymentMessage =
  state == 'success'
    ? 1 //'Payment completed successfully'

  : state == 'processing'
    ? 2 //'Payment processing'

  : state == 'invalid_cvc'
    ? 3 //'There was an issue with your CVC number'

  : true //state == 'invalid_expiry'
    ? 4 //'Expiry must be sometime in the past.'

    : '5 // There was an issue with the payment.  Please contact support.'
```

It's possible to do that with operator last, but it's trickier.

However this lands I'm really excited and grateful for ternaries getting the attention they deserve.

**Comment by mysticatea** [comment]
Hi. I'd like to vote the alternate ternary formatting, especially for TypeScript.

TypeScript introduced Conditional Types: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html

```ts
// An official sample code
type TypeName<T> =
    T extends string ? "string" :
    T extends number ? "number" :
    T extends boolean ? "boolean" :
    T extends undefined ? "undefined" :
    T extends Function ? "function" :
    "object";
```

Prettier's current format is not nice for that in my 2 cents:

```ts
// Prettier does
type TypeName<T> = T extends string
    ? "string"
    : T extends number
        ? "number"
        : T extends boolean
            ? "boolean"
            : T extends undefined
                ? "undefined"
                : T extends Function ? "function" : "object"
```

----

FWIW, ESLint's indent rule has the option for the alternate ternary formatting: `flatTernaryExpressions` in https://eslint.org/docs/rules/indent#options

**Comment by jessejanderson** [comment]
While there are a lot of differing opinions in here, it seems like prettier could at least support eslint's `flatTernaryExpressions` rule that @mysticatea linked to.  This would would help for those (myself included) that prefer the style @mvolkmann illustrated at the start this issue.

**Comment by marksamman** [comment]
The formatting of ternary operators is the only reason my organization is hesitant to use prettier. The excessive nesting affects readability negatively for us. I hope you'll at least consider adding an option with less nesting if the default behavior isn't changed.

Here's an example (our preferred style) of how we currently format ternary operators:
```jsx
<div className={'match-achievement-medal-type type' + (medals[0].record ? '-record' : (medals[0].unique ? '-unique' : medals[0].type))}>
	{medals[0].record ? (
		i18n('Record')
	) : medals[0].unique ? (
		i18n('Unique')
	) : medals[0].type === 0 ? (
		i18n('Silver')
	) : medals[0].type === 1 ? (
		i18n('Gold')
	) : medals[0].type === 2 ? (
		i18n('Platinum')
	) : (
		i18n('Theme')
	)}
</div>
```

This is what prettier turns it into:
```jsx
<div
	className={
		"match-achievement-medal-type type" +
		(medals[0].record
			? "-record"
			: medals[0].unique
				? "-unique"
				: medals[0].type)
	}
>
	{medals[0].record
		? i18n("Record")
		: medals[0].unique
			? i18n("Unique")
			: medals[0].type === 0
				? i18n("Silver")
				: medals[0].type === 1
					? i18n("Gold")
					: medals[0].type === 2
						? i18n("Platinum")
						: i18n("Theme")}
</div>
```

**Comment by j-f1** [comment]
How about switching to an alternate style if there are three or more nested ternaries?

**Comment by lydell** [comment]
@j-f1 Interesting idea! I think it might make sense to flatten ternaries, just like we flatten other operators such as sequences of `+`.

**Comment by marksamman** [comment]
Would you be open to a pull request that adds a configuration option with a different style? If so, is there a style that doesn't use additional nesting that you would approve?

**Comment by lydell** [comment]
@marksamman We're more interested in improving/tweaking/changing the way Prettier prints ternaries rather adding configuration.

**Comment by jessejanderson** [comment]
@lydell what about supporting the eslint `flatTernaryExpressions` rule, which would address this situation without needing additional configuration inside of Prettier itself?

**Comment by lydell** [comment]
A proof-of-concept PR would be appreciated, but I can't guarantee that it will be merged. :+1:

**Comment by zeorin** [comment]
@lydell @suchipi I've created a PR for nested ternaries: #4767.

**Comment by j-f1** [comment]
Adding @suchipi due to https://github.com/prettier/prettier/pull/4767#issuecomment-414031218.

**Comment by suchipi** [comment]
Thanks, I accidentally assigned myself to the PR instead of the issue.

**Comment by SelaO** [comment]
How can I override this terribly unreadable setting? 

I don't want to place every part of my nested ternaries in a new line either.

**Comment by suchipi** [comment]
@SelaO for now, `// prettier-ignore` comments or using prettier in conjunction with some ESLint rule are the only workarounds.

**Comment by lydell** [comment]
Oops, sorry I forgot about that one!

**Comment by rattrayalex** [comment]
This was closed by #5039, which is likely to be reverted in #9559, so I have resumed this discussion about alternative formats for nested ternaries in #9561 – feedback welcome there!
